Makefile

```diff
@@ -1,137 +1,205 @@
-PHP_CS_FIXER_FUTURE_MODE=1
-PHPSTAN=./phpstan.phar
-PHP-CS-FIXER=./php-cs-fixer-v2.phar
-PHPUNIT=vendor/bin/phpunit
-INFECTION=build/bin/infection.phar
+.DEFAULT_GOAL := help
+
+.PHONY: help
+# TODO: add a check to ensure the result is always the expected one
+# TODO: add a check to ensure whenever a PHONY is declared, the following target has the same name; e.g. here .PHONY=help
+# so the next target should be `help:`. Otherwise this is most certainly a mistake.
+help:
+	@echo "\033[33mUsage:\033[0m\n  make TARGET\n\n\033[32m#\n# Commands\n#---------------------------------------------------------------------------\033[0m\n"
+	@fgrep -h "##" $(MAKEFILE_LIST) | fgrep -v fgrep | sed -e 's/\\$$//' | sed -e 's/##//' | awk 'BEGIN {FS = ":"}; {printf "\033[33m%s:\033[0m%s\n", $$1, $$2}'
+
 
-# URLs to download all tools
-BOX_URL="https://github.com/humbug/box/releases/download/3.1.0/box.phar"
-PHP-CS-FIXER_URL="https://cs.sensiolabs.org/download/php-cs-fixer-v2.phar"
+#
+# Variables
+#---------------------------------------------------------------------------
+BOX=./.tools/box
+BOX_URL="https://github.com/humbug/box/releases/download/3.8.0/box.phar"
+
+PHP_CS_FIXER=./.tools/php-cs-fixer
+PHP_CS_FIXER_URL="https://cs.sensiolabs.org/download/php-cs-fixer-v2.phar"
+PHP_CS_FIXER_CACHE=build/cache/.php_cs.cache
+
+PHPSTAN=./.tools/phpstan
 PHPSTAN_URL="https://github.com/phpstan/phpstan/releases/download/0.11.5/phpstan.phar"
 
-FLOCK=./devTools/flock
+PHPUNIT=vendor/bin/phpunit
+
+INFECTION=./build/infection.phar
 
-DOCKER_RUN=docker run -t --rm -v "$$PWD":/opt/infection -w /opt/infection
+DOCKER_RUN=docker run --tty --rm --volume "$$PWD":/opt/infection --workdir /opt/infection
 DOCKER_RUN_72=$(FLOCK) devTools/*php72*.json $(DOCKER_RUN) infection_php72
+DOCKER_RUN_72_IMAGE=devTools/Dockerfile-php72-xdebug.json
 DOCKER_RUN_73=$(FLOCK) devTools/*php73*.json $(DOCKER_RUN) infection_php73
+DOCKER_RUN_73_IMAGE=devTools/Dockerfile-php73-xdebug.json
 
-.PHONY: all
-#Run all checks, default when running 'make'
-all: analyze test
+FLOCK=./devTools/flock
 
-#Non phony targets for phars etc.
-vendor: composer.json composer.lock
-	composer install
 
-build/cache:
-	mkdir -p build/cache
+#
+# Commands (phony targets)
+#---------------------------------------------------------------------------
 
-./php-cs-fixer-v2.phar:
-	wget $(PHP-CS-FIXER_URL)
-	chmod a+x ./php-cs-fixer-v2.phar
+.PHONY: compile
+compile:	## Bundles Infection into a PHAR
+compile: $(INFECTION)
 
-./phpstan.phar:
-	wget $(PHPSTAN_URL)
-	chmod a+x ./phpstan.phar
+.PHONY: cs-fix
+cs-fix:	  	## Runs PHP-CS-Fixer
+cs-fix: $(PHP_CS_FIXER)
+	# TODO: disable verbose mode?
```

Comment:
Should the verbose mode be disabled for PHP-CS-Fixer? What's the point of having it if it's run only locally? IMO if you want extra verbosity you can run the command manually instead

Reply:
It depends on the situation with files with syntax errors. I had some problem with PHP-CS-Fixer silently ignoring such files in the default non-verbose mode, but this may be the thing of past.

(I prefer to run PHP-CS-Fixer in the verbose mode at all times because it may go great lengths looking over things do not need fixing if you happen to, say, make a very unfortunate symlink somewhere else. At least you will see it being stuck. It is not like Infection needs any of the symlinks, just saying.)

Reply:
ok